text/0123-pending-code-as-storage-location-for-runtime-upgrades.md

```diff
@@ -0,0 +1,60 @@
+# RFC-0123: Introduce `:pending_code` as intermediate storage key for the runtime code
+
+| **Start Date**  | 14.10.2024                                                              |
+| --------------- | ----------------------------------------------------------------------- |
+| **Description** | Store a runtime upgrade in `:pending_code` before moving it to `:code`. |
+| **Authors**     | Bastian Köcher                                                          |
+
+## Summary
+
+The code of a runtime is stored in its own state, and when performing a runtime upgrade, this code is replaced. The new runtime can contain runtime migrations that adapt the state to the state layout as defined by the runtime code. This runtime migration is executed when building the first block with the new runtime code. Anything that interacts with the runtime state uses the state layout as defined by the runtime code. So, when trying to load something from the state in the block that applied the runtime upgrade, it will use the new state layout but will decode the data from the non-migrated state. In the worst case, the data is incorrectly decoded, which may lead to crashes or halting of the chain.
+
+This RFC proposes to store the new runtime code under a different storage key when applying a runtime upgrade. This way, all the off-chain logic can still load the old runtime code under the default storage key and decode the state correctly. The block producer is then required to use this new runtime code to build the next block. While building the next block, the runtime is executing the migrations and moves the new runtime code to the default runtime code location. So, the runtime code found under the default location is always the correct one to decode the state from which the runtime code was loaded.
```

Comment:
What if we need multi block migrations?

Reply:
https://github.com/polkadot-fellows/RFCs/pull/123/files#diff-249dca18be7886ab8cd41ab977966230ddcde054f96cd16571e86af54f9e7d02R30